### PR TITLE
Update to compatible with aiida-core 2.x

### PR DESCRIPTION
**Pull request by unkcpz**
Hi @greschd @bosonie,

Here are some attempts to remove the dependency of `aiida-pytest`. All test for `run` passed but not for `submit` caused by issue I opened at https://github.com/aiidateam/aiida-core/issues/5511. Do you have better suggestion on how to do that?

Then, since we drop 1.x, other thing I can image to apply here is that we can use dot `.` to access and construct namespace instead of double score `__` for `nested_links`. Whatelse any new features you think we can update and implemented for this tool?

I can also manage to update the pre-commit dependencies and release workflow if you prefer, I can open another PR for that.

### CODE REVIEW

File: setup.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
   ],
   "reentry_register": true,
   "install_requires": [
-    "aiida-core>=1.3.0<2.0.0",
+    "aiida-core>=2.0.0",
```
Comment (greschd) [inline setup.json]: Because we must assume that `3.x` will again be backwards-incompatible (otherwise it probably wouldn't be a major release), the version constraint here should be `>=2.0.0<3.0.0`.

**Review by greschd — COMMENTED** [review]
> Then, since we drop 1.x, other thing I can image to apply here is that we can use dot . to access and construct namespace instead of double score __ for nested_links.

Yeah, I think that'd make sense. Also agree on updating the pre-commit hooks etc.

As for the `submit` tests: I'm not entirely sure what's different between this code and what I was doing in `aiida-pytest`; maybe it's the `AIIDA_PATH` environment variable that needs to be set?

https://github.com/greschd/aiida_pytest/blob/823d8066fd37288132e31314de511c2cf991538d/aiida_pytest/_configure.py#L107

As for the necessity of running the test with both `run` and `submit` at all.. I'm not entirely sure if this is still needed. There used to be subtle differences where some things couldn't get serialized properly on `submit`, but this is just as much a test of `aiida-core` functionality as it tests `aiida-optimize`.
TL;DR: If it's simple enough to keep the `submit` tests around I'm all for it, but if that ends up being the major blocker I would be open to removing them.

File: setup.json [hunk 2]
```diff
@@ -29,7 +29,7 @@
   ],
   "reentry_register": true,
```
Comment (greschd) [inline setup.json]: Could it be that this is no longer needed?

Comment (unkcpz) [reply setup.json]: Indeed!

**Comment by unkcpz** [comment]
> TL;DR: If it's simple enough to keep the submit tests around I'm all for it, but if that ends up being the major blocker I would be open to removing them.

@greschd thanks! I'll wait a bit too see whether it is easy for https://github.com/aiidateam/aiida-core/issues/5511 to be fixed. Then we just need to add `submit` option back to fixture. 

EDIT: could you approve the running of CI test? thanks.

**Comment by greschd** [comment]
The ` --quiet-wipe --print-status` are flags from `aiida-pytest`, those should be removed in the pipeline.

**Comment by unkcpz** [comment]
I borrow the concept from `aiida-core`, start a daemon and submit the process there for testing. But only running one submit test there for testing purposes. If you are good with this change I will add at least one test for every engine and wrapper.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -20,7 +20,7 @@ jobs:
       - name: Install python dependencies
         env:
           INSTALL_TYPE: docs
```
Comment (greschd) [inline .github/workflows/ci.yml]: I think the `INSTALL_TYPE` is now unused, right?

**Comment by greschd** [comment]
> I borrow the concept from `aiida-core`, start a daemon and submit the process there for testing. But only running one submit test there for testing purposes. If you are good with this change I will add at least one test for every engine and wrapper.

Yep, makes sense :+1:

**Comment by bosonie** [comment]
Hi there! What is the status of this? 
Tests on python 3.7 must be removed since aiida 2.0.0 does not support it anymore.
Please let me know if I can help somehow. Otherwise a release after this merge will be appreciated, so I can subsequently release the `aiida-siesta` version compatible with 2.0. Thanks!

**Comment by unkcpz** [comment]
Hi @bosonie, I think this PR is almost ready. I didn't get time to manage to add daemon tests for every engine last week, but I think it is okay to add them in a separate PR? Meanwhile, I want to leave cleanup double underscore `__` into a separate PR, since I make some attempts and I think using `__` to access the node as it stores is more clear for now. 
I'll remove `INSTALL_TYPE` as @greschd mentioned and we can try to merge this.

File: .github/system_tests/sample_processes.py [hunk 4]
```diff
@@ -0,0 +1 @@
+../../tests/sample_processes.py
```
Comment (greschd) [inline .github/system_tests/sample_processes.py:1]: Uhmm.. how does this work?

File: .github/system_tests/test_daemon.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        ))
+    except subprocess.CalledProcessError as exception:
+        print(f'Note: the command failed, message: {exception}')
+
+
+def jobs_have_finished(pks):
+    """Check if jobs with given pks have finished."""
+    finished_list = [load_node(pk).is_terminated for pk in pks]
+    node_list = [load_node(pk) for pk in pks]
+    num_finished = len([_ for _ in finished_list if _])
```
Comment (greschd) [inline .github/system_tests/test_daemon.py]: Nitpick: `sum(finished_list)`?

File: .github/system_tests/test_daemon.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    for node in node_list:
+        if not node.is_terminated:
+            print(f'not terminated: {node.pk} [{node.process_state}]')
+    print(f'{num_finished}/{len(finished_list)} finished')
+    return False not in finished_list
```
Comment (greschd) [inline .github/system_tests/test_daemon.py]: `return all(finished_list)` is clearer IMO

File: .github/system_tests/test_daemon.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    try:
+        print(subprocess.check_output(
+            ['cat', f'{daemon_log}'],
+            stderr=subprocess.STDOUT,
+        ))
+    except subprocess.CalledProcessError as exception:
+        print(f'Note: the command failed, message: {exception}')
+
+
+def jobs_have_finished(pks):
```
Comment (greschd) [inline .github/system_tests/test_daemon.py]: Hmm wait.. is this function used anywhere?

File: .github/workflows/nightly.yml [hunk 8]
```diff
@@ -0,0 +1,62 @@
+name: nightly
```
Comment (greschd) [inline .github/workflows/nightly.yml]: So these are _just_ the daemon tests, correct? I'm not sure they warrant a separate pipeline, or if they should be a step in the "regular" CI.

Since this is a much smaller project than `aiida-core`, I'd like to keep the testing infrastructure at a smaller level also.

File: aiida_optimize/engines/_particle_swarm.py [hunk 9]
```diff
@@ -110,8 +110,8 @@ def submit_initialize(self):
         self.fun_global_best = np.inf
         # Initialize the velocities to random number in [-1,1]
         self.velocities = np.zeros((n_parts, n_vars))
-        for line in range(len(self.velocities)):
-            for col in range(len(self.velocities[line])):
+        for line in range(len(self.velocities)):  # pylint: disable=consider-using-enumerate
```
Comment (greschd) [inline aiida_optimize/engines/_particle_swarm.py]: Hmm, I tend to agree with `pylint` here: `for line, _ in enumerate(self.velocities)` looks nicer to me. But both are fine, if you feel strongly about it.

File: tests/wrappers/test_add_inputs_wrapper.py [hunk 10]
```diff
@@ -2,18 +2,19 @@
...
 from aiida.engine import run_get_node
 
 from aiida_optimize.wrappers import AddInputsWorkChain
 
-from sample_processes import echo_process, EchoDictValue, EchoNestedValues  # pylint: disable=import-error,useless-suppression, unused-import
+from sample_processes import EchoDictValue, EchoNestedValues  # pylint: disable=import-error,useless-suppression, unused-import
 
 
-def test_basic(configure_with_daemon, echo_process):
+@pytest.mark.usefixtures('aiida_profile_clean')
```
Comment (greschd) [inline tests/wrappers/test_add_inputs_wrapper.py:16]: Can we `autouse` this fixture?

**Review by greschd — COMMENTED** [review]
Thanks for all your work @unkcpz!

Sorry for being a bit slow to review, I've been out sick a few days.

Anyways, I've got mostly some nitpicks / questions. In cases where you disagree, feel free to let me know I'm wrong :smile:

File: aiida_optimize/engines/_particle_swarm.py (hunk 9 above)
Comment (unkcpz) [reply aiida_optimize/engines/_particle_swarm.py]: Agree.

File: .github/workflows/nightly.yml (hunk 8 above)
Comment (unkcpz) [reply .github/workflows/nightly.yml]: Fair point, I move it to ci.yml.

File: .github/system_tests/test_daemon.py (hunk 7 above)
Comment (unkcpz) [reply .github/system_tests/test_daemon.py]: Yes, it is not used at last. Purged.

File: .github/system_tests/sample_processes.py (hunk 4 above)
Comment (unkcpz) [reply .github/system_tests/sample_processes.py:1]: It is simply a soft link to the `sample_processes.py` 😆. Not sure it is clear, but otherwise have to copy and paste code from there.

File: .github/system_tests/test_daemon.py (hunk 5 above)
Comment (unkcpz) [reply .github/system_tests/test_daemon.py]: I personally don't quite a fan of sum a bool list. But fine to change if you prefer.

File: tests/wrappers/test_add_inputs_wrapper.py (hunk 10 above)
Comment (unkcpz) [reply tests/wrappers/test_add_inputs_wrapper.py:16]: I think only tests that involve actually running the calculation need to clean the profile after finished.

**Comment by unkcpz** [comment]
> Sorry for being a bit slow to review, I've been out sick a few days.
Anyways, I've got mostly some nitpicks / questions. In cases where you disagree, feel free to let me know I'm wrong 😄

No worries, I hope you are feeling better. I am also busy these days, so sorry for not being able to manage to add more submit tests and implement all things nicely. 

I update the PR, hope everything is clear.

File: .github/system_tests/sample_processes.py (hunk 4 above)
Comment (greschd) [reply .github/system_tests/sample_processes.py:1]: Ah yeah that's fine, I guess I was puzzled by the rendering online.

File: .github/system_tests/test_daemon.py (hunk 6 above)
Comment (unkcpz) [reply .github/system_tests/test_daemon.py]: The method is not used as you mentioned below.

File: .github/system_tests/test_daemon.py (hunk 5 above)
Comment (greschd) [reply .github/system_tests/test_daemon.py]: Nah, it's fine to leave as-is.

**Comment by greschd** [comment]
Aand merged! :tada:

Thanks again @unkcpz!
